// @rpldy/shared@1.10.0/lib/cjs/utils/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "clone", {
  enumerable: true,
  get: function () {
    return _clone.default;
  }
});
Object.defineProperty(exports, "devFreeze", {
  enumerable: true,
  get: function () {
    return _devFreeze.default;
  }
});
Object.defineProperty(exports, "getMerge", {
  enumerable: true,
  get: function () {
    return _merge.getMerge;
  }
});
Object.defineProperty(exports, "hasWindow", {
  enumerable: true,
  get: function () {
    return _hasWindow.default;
  }
});
Object.defineProperty(exports, "isEmpty", {
  enumerable: true,
  get: function () {
    return _isEmpty.default;
  }
});
Object.defineProperty(exports, "isFunction", {
  enumerable: true,
  get: function () {
    return _isFunction.default;
  }
});
Object.defineProperty(exports, "isPlainObject", {
  enumerable: true,
  get: function () {
    return _isPlainObject.default;
  }
});
Object.defineProperty(exports, "isProduction", {
  enumerable: true,
  get: function () {
    return _isProduction.default;
  }
});
Object.defineProperty(exports, "isPromise", {
  enumerable: true,
  get: function () {
    return _isPromise.default;
  }
});
Object.defineProperty(exports, "isSamePropInArrays", {
  enumerable: true,
  get: function () {
    return _isSamePropInArrays.default;
  }
});
Object.defineProperty(exports, "merge", {
  enumerable: true,
  get: function () {
    return _merge.default;
  }
});
Object.defineProperty(exports, "pick", {
  enumerable: true,
  get: function () {
    return _pick.default;
  }
});
Object.defineProperty(exports, "scheduleIdleWork", {
  enumerable: true,
  get: function () {
    return _scheduleIdleWork.default;
  }
});
var _isFunction = _interopRequireDefault(require("./isFunction"));
var _isSamePropInArrays = _interopRequireDefault(require("./isSamePropInArrays"));
var _devFreeze = _interopRequireDefault(require("./devFreeze"));
var _merge = _interopRequireWildcard(require("./merge"));
var _clone = _interopRequireDefault(require("./clone"));
var _pick = _interopRequireDefault(require("./pick"));
var _isPlainObject = _interopRequireDefault(require("./isPlainObject"));
var _hasWindow = _interopRequireDefault(require("./hasWindow"));
var _isProduction = _interopRequireDefault(require("./isProduction"));
var _isPromise = _interopRequireDefault(require("./isPromise"));
var _scheduleIdleWork = _interopRequireDefault(require("./scheduleIdleWork"));
var _isEmpty = _interopRequireDefault(require("./isEmpty"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
