// @suspensive/react@2.18.12/dist/index.js
"use client"
import {
  wrap
} from "./chunk-LXXAF22O.js";
import {
  Delay
} from "./chunk-YARTG4UY.js";
import {
  DevMode
} from "./chunk-CVLUUUEY.js";
import {
  ErrorBoundary,
  useErrorBoundary,
  useErrorBoundaryFallbackProps
} from "./chunk-LVRP5XPR.js";
import {
  ErrorBoundaryGroup,
  useErrorBoundaryGroup
} from "./chunk-S5J6B5JD.js";
import {
  Suspense
} from "./chunk-H25YTWH6.js";
import "./chunk-ZJULBOVW.js";
import {
  ClientOnly
} from "./chunk-PXLOC2DW.js";
import "./chunk-XMHMZEKB.js";
import {
  Suspensive,
  SuspensiveProvider
} from "./chunk-3HSSVKIT.js";
import {
  DefaultProps,
  DefaultPropsProvider
} from "./chunk-U7CK4ZVL.js";
import "./chunk-NF2UFYHD.js";
import "./chunk-PJ6VIZI5.js";
import "./chunk-HVDFXOFP.js";
export {
  ClientOnly,
  DefaultProps,
  DefaultPropsProvider,
  Delay,
  DevMode,
  ErrorBoundary,
  ErrorBoundaryGroup,
  Suspense,
  Suspensive,
  SuspensiveProvider,
  useErrorBoundary,
  useErrorBoundaryFallbackProps,
  useErrorBoundaryGroup,
  wrap
};
