// @rpldy/retry@1.10.0/lib/cjs/retry.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _shared = require("@rpldy/shared");
var _simpleState = _interopRequireDefault(require("@rpldy/simple-state"));
var _uploader = require("@rpldy/uploader");
var _consts = require("./consts");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const FAILED_STATES = [_shared.FILE_STATES.ABORTED, _shared.FILE_STATES.ERROR];
const removeItemFromState = (retryState, id) => {
  retryState.updateState(state => {
    const batchId = state.failed[id].batchId;
    delete state.failed[id];
    const biMapIndex = state.batchIdsMap[batchId].indexOf(id);
    state.batchIdsMap[batchId].splice(biMapIndex, 1);
    if (!state.batchIdsMap[batchId].length) {
      delete state.batchIdsMap[batchId];
    }
  });
};
const uploadFailedIds = (uploader, retryState, trigger, ids, options) => {
  const failed = retryState.getState().failed;
  const failedIds = ids || Object.keys(failed);
  const uploads = failedIds.map(id => failed[id]).filter(Boolean);
  if (uploads.length) {
    const retryOptions = {
      ...(options || null),
      autoUpload: typeof options?.autoUpload !== "undefined" ? options.autoUpload : true
    };
    trigger(_consts.RETRY_EVENT, {
      items: uploads,
      options: retryOptions
    });
    failedIds?.forEach(id => removeItemFromState(retryState, id));
    uploader.add(uploads, retryOptions);
  }
  return !!uploads.length;
};
const retryItem = (uploader, retryState, trigger, itemId, options) => {
  _shared.logger.debugLog(`uploady.retry: about to retry item: ${itemId}`);
  return uploadFailedIds(uploader, retryState, trigger, [itemId], options);
};
const retry = (uploader, retryState, trigger, itemId, options) => {
  let result;
  if (itemId) {
    result = retryItem(uploader, retryState, trigger, itemId, options);
  } else {
    _shared.logger.debugLog(`uploady.retry: about to retry all failed item`);
    result = uploadFailedIds(uploader, retryState, trigger, null, options);
  }
  return result;
};
const retryBatch = (uploader, retryState, trigger, batchId, options) => {
  _shared.logger.debugLog(`uploady.retry: about to retry batch: ${batchId}`);
  const batchItemIds = retryState.getState().batchIdsMap[batchId]?.slice();
  return batchItemIds ? uploadFailedIds(uploader, retryState, trigger, batchItemIds, options) : false;
};
const createRetryState = () => {
  const {
    state,
    update
  } = (0, _simpleState.default)({
    batchIdsMap: {},
    failed: {}
  });
  return {
    updateState: updater => {
      update(updater);
    },
    getState: () => state
  };
};
const registerEvents = (uploader, retryState) => {
  uploader.on(_uploader.UPLOADER_EVENTS.ITEM_FINALIZE, item => {
    if (FAILED_STATES.includes(item.state)) {
      retryState.updateState(state => {
        state.failed[item.id] = item;
        const biMap = state.batchIdsMap[item.batchId] = state.batchIdsMap[item.batchId] || [];
        biMap.push(item.id);
      });
    }
  });
};
const retryEnhancer = (uploader, trigger) => {
  const retryState = createRetryState();
  registerEvents(uploader, retryState);
  uploader.registerExtension(_consts.RETRY_EXT, {
    retry: (itemId, options) => retry(uploader, retryState, trigger, itemId, options),
    retryBatch: (batchId, options) => retryBatch(uploader, retryState, trigger, batchId, options)
  });
  return uploader;
};
var _default = exports.default = retryEnhancer;
