// @rpldy/uploader@1.10.0/lib/cjs/consts.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UPLOADER_EVENTS = exports.SENDER_EVENTS = exports.PROGRESS_DELAY = exports.ITEM_FINALIZE_STATES = void 0;
var _shared = require("@rpldy/shared");
const UPLOADER_EVENTS = exports.UPLOADER_EVENTS = (0, _shared.devFreeze)({
  BATCH_ADD: "BATCH-ADD",
  BATCH_START: "BATCH-START",
  BATCH_PROGRESS: "BATCH_PROGRESS",
  BATCH_FINISH: "BATCH-FINISH",
  BATCH_ABORT: "BATCH-ABORT",
  BATCH_CANCEL: "BATCH-CANCEL",
  BATCH_ERROR: "BATCH-ERROR",
  BATCH_FINALIZE: "BATCH-FINALIZE",
  ITEM_START: "FILE-START",
  ITEM_CANCEL: "FILE-CANCEL",
  ITEM_PROGRESS: "FILE-PROGRESS",
  ITEM_FINISH: "FILE-FINISH",
  ITEM_ABORT: "FILE-ABORT",
  ITEM_ERROR: "FILE-ERROR",
  ITEM_FINALIZE: "FILE-FINALIZE",
  REQUEST_PRE_SEND: "REQUEST_PRE_SEND",
  ALL_ABORT: "ALL_ABORT"
});
const PROGRESS_DELAY = exports.PROGRESS_DELAY = 50;
const SENDER_EVENTS = exports.SENDER_EVENTS = (0, _shared.devFreeze)({
  ITEM_PROGRESS: "ITEM_PROGRESS",
  BATCH_PROGRESS: "BATCH_PROGRESS"
});
const ITEM_FINALIZE_STATES = exports.ITEM_FINALIZE_STATES = [_shared.FILE_STATES.FINISHED, _shared.FILE_STATES.ERROR, _shared.FILE_STATES.CANCELLED, _shared.FILE_STATES.ABORTED];
