// @rpldy/shared@1.10.0/lib/cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "BATCH_STATES", {
  enumerable: true,
  get: function () {
    return _consts.BATCH_STATES;
  }
});
Object.defineProperty(exports, "FILE_STATES", {
  enumerable: true,
  get: function () {
    return _consts.FILE_STATES;
  }
});
Object.defineProperty(exports, "XhrPromise", {
  enumerable: true,
  get: function () {
    return _request.XhrPromise;
  }
});
Object.defineProperty(exports, "clone", {
  enumerable: true,
  get: function () {
    return _utils.clone;
  }
});
Object.defineProperty(exports, "createBatchItem", {
  enumerable: true,
  get: function () {
    return _batchItem.default;
  }
});
Object.defineProperty(exports, "devFreeze", {
  enumerable: true,
  get: function () {
    return _utils.devFreeze;
  }
});
Object.defineProperty(exports, "getIsBatchItem", {
  enumerable: true,
  get: function () {
    return _batchItem.getIsBatchItem;
  }
});
Object.defineProperty(exports, "getMerge", {
  enumerable: true,
  get: function () {
    return _utils.getMerge;
  }
});
Object.defineProperty(exports, "hasWindow", {
  enumerable: true,
  get: function () {
    return _utils.hasWindow;
  }
});
Object.defineProperty(exports, "invariant", {
  enumerable: true,
  get: function () {
    return _invariant.default;
  }
});
Object.defineProperty(exports, "isEmpty", {
  enumerable: true,
  get: function () {
    return _utils.isEmpty;
  }
});
Object.defineProperty(exports, "isFunction", {
  enumerable: true,
  get: function () {
    return _utils.isFunction;
  }
});
Object.defineProperty(exports, "isPlainObject", {
  enumerable: true,
  get: function () {
    return _utils.isPlainObject;
  }
});
Object.defineProperty(exports, "isProduction", {
  enumerable: true,
  get: function () {
    return _utils.isProduction;
  }
});
Object.defineProperty(exports, "isPromise", {
  enumerable: true,
  get: function () {
    return _utils.isPromise;
  }
});
Object.defineProperty(exports, "isSamePropInArrays", {
  enumerable: true,
  get: function () {
    return _utils.isSamePropInArrays;
  }
});
exports.logger = void 0;
Object.defineProperty(exports, "merge", {
  enumerable: true,
  get: function () {
    return _utils.merge;
  }
});
Object.defineProperty(exports, "parseResponseHeaders", {
  enumerable: true,
  get: function () {
    return _request.parseResponseHeaders;
  }
});
Object.defineProperty(exports, "pick", {
  enumerable: true,
  get: function () {
    return _utils.pick;
  }
});
Object.defineProperty(exports, "request", {
  enumerable: true,
  get: function () {
    return _request.default;
  }
});
Object.defineProperty(exports, "scheduleIdleWork", {
  enumerable: true,
  get: function () {
    return _utils.scheduleIdleWork;
  }
});
Object.defineProperty(exports, "throttle", {
  enumerable: true,
  get: function () {
    return _justThrottle.default;
  }
});
Object.defineProperty(exports, "triggerCancellable", {
  enumerable: true,
  get: function () {
    return _triggerCancellable.default;
  }
});
Object.defineProperty(exports, "triggerUpdater", {
  enumerable: true,
  get: function () {
    return _triggerUpdater.default;
  }
});
var _invariant = _interopRequireDefault(require("invariant"));
var _justThrottle = _interopRequireDefault(require("just-throttle"));
var _consts = require("./consts");
var logger = _interopRequireWildcard(require("./logger"));
exports.logger = logger;
var _triggerCancellable = _interopRequireDefault(require("./triggerCancellable"));
var _triggerUpdater = _interopRequireDefault(require("./triggerUpdater"));
var _batchItem = _interopRequireWildcard(require("./batchItem"));
var _request = _interopRequireWildcard(require("./request"));
var _utils = require("./utils");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
